packages/repository/src/query.ts

```diff
@@ -688,6 +698,72 @@ export class FilterBuilder<MT extends object = AnyObject> {
   }
 }
 
+/**
+ * Checks whether fields array passed as an argument is a
+ * subset of fields picked by a target filter.
+ *
+ * @param fields - An array of fields to search in a filter
+ * @param filter - A target filter
+ */
+export function matchesFields<T extends object = AnyObject>(
+  fields: (keyof T)[],
+  filter?: Filter<T>,
+) {
+  const normalized = new FilterBuilder(filter).build();
+  const targetFields = normalized.fields;
+  if (!targetFields) {
+    return true;
+  }
+  const isDisablingOnly =
+    _.size(targetFields) > 0 && !_.some(targetFields, Boolean);
+  for (const f of fields) {
+    if (!targetFields[f] && (f in targetFields || !isDisablingOnly)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * Ensures that queries which apply the returned filter would always include
+ * the target fields. To undo this effect later, fields that were disabled
+ * in the original filter will be added to the pruning mask.
+ *
+ * @param targetFields - An array of fields to include
+ * @param filter - A target filter
+ * @returns A tuple containing amended filter and pruning mask
+ */
+export function ensureFields<T extends object = AnyObject>(
+  targetFields: (keyof T)[],
+  filter: Filter<T>,
+) {
+  const builder = new FilterBuilder(filter);
+  const fields = builder.build().fields;
+  if (!fields || matchesFields(targetFields, filter)) {
+    return [filter, {} as PruningMask<T>] as const;
+  }
+  const isDisablingOnly = _.size(fields) > 0 && !_.some(fields, Boolean);
+  const prune = (isDisablingOnly ? _.keys(fields) : []) as (keyof T)[];
+  targetFields.forEach(f => {
+    if (!fields[f]) {
+      prune.push(f);
+      builder.fields(f);
+    }
+  });
+  const pruneFields = _.fromPairs(
+    prune.map(f => [f, undefined]),
+  ) as PruningMask<T>;
+
+  const newFilter = builder.build();
+  // if the filter only hides the fields, unset the entire fields clause
+  if (isDisablingOnly) {
+    delete newFilter.fields;
+  }
+  return [newFilter, pruneFields] as const;
```

Comment:
Let's return an object with named properties instead of tuple please, that way it's easier to understand what values are returned by this helper. See e.g. one of the older versions of your patch (e29027560):

```ts
  return {
    filter: builder.build(),
    pruneFields,
  };
```

Reply:
(But please choose a different property name than `pruneFields`, see https://github.com/strongloop/loopback-next/pull/5592#issuecomment-646665314):

> The method `ensureFields` is modifying the query to ensure given fields are always included. I think we should not limit this method to inclusion resolving only (where we want to prune those fields afterwards) and we should think about it in more generic terms. Instead of returning `pruneFields` object that's specific to inclusion resolvers, I'd prefer to return a list of field names that were added to the query on top of what was requested by the user. Example names: `fieldsAdded`, `extraFields`, etc.

Reply:
Fixed